File: packages/python/qsu/math/round.py
from decimal import Decimal, ROUND_HALF_UP

from ._decimalShift import _decimalShift


def round(value, precision=0):
	# `Decimal(str(value))` reads the shortest representation of the float, which is the
	# decimal the caller wrote. `Decimal(value)` would read the binary double instead, so
	# `1.005` would arrive as `1.00499999999999989...` and round down to `1`.
	number = Decimal(str(value))

	if not number.is_finite():
		return value

	# `ROUND_HALF_UP` in `decimal` means ties away from zero, unlike the built-in `round`,
	# which uses banker's rounding and answers `0` for `0.5` and `2` for `2.5`.
	rounded = _decimalShift(number, precision).to_integral_value(rounding=ROUND_HALF_UP)
	result = _decimalShift(rounded, -precision)

	return int(result) if result == result.to_integral_value() else float(result)

File: packages/javascript/lib/math/round.ts
import { decimalShift } from './_decimalShift.js';

export function round(value: number, precision = 0): number {
	if (!Number.isFinite(value)) {
		return value;
	}

	const shifted = decimalShift(value, precision);
	// Ties go away from zero. `Math.round` sends them toward positive infinity, so `-0.5`
	// would answer `-0` here where Dart and Python answer `-1`.
	const rounded = shifted < 0 ? -Math.round(-shifted) : Math.round(shifted);

	return decimalShift(rounded, -precision);
}
